skrifa/src/color/color.rs

```diff
@@ -0,0 +1,259 @@
+
+use crate::prelude::LocationRef;
+
+#[derive(Clone, Debug, Default)]
+// This repr(C) is required so that C-side FFI's
+// are able to cast the ColorStop slice to a C-side array pointer.
+#[repr(C)]
+pub struct ColorStop {
+    pub offset: f32,
+    pub palette_index: u16,
+    pub alpha: f32,
+}
+
+// Design considerations: In principle, a local Vec allocation is not required if we're willing
+// to walk the ResolvedColorStop iterator find and the minimum and maximum color stops.
+// Then we could scale the color stops based on those. But that would require re-applying the deltas
+// at least once, just to pass it to the client side and have the client sort the
+// collected items once again. But then if we use an Iterator<Item=ColorStop> instead as the
+// color_stops field to pass the color stops, then we would need a Fontations-side allocations
+// to sort, and an extra allocation on the client side to .collect().
+#[derive(Debug)]
+pub enum Brush<'a> {
+    Solid {
+        palette_index: u16,
+        alpha: f32,
+    },
+    // Normalized to a straight line between points p0 and p1,
+    // color stops normalized to align with both points.
+    LinearGradient {
+        p0: Point<f32>,
+        p1: Point<f32>,
+        color_stops: &'a [ColorStop],
+        extend: u8,
+    },
+    RadialGradient {
+        c0: Point<f32>,
+        r0: f32,
+        c1: Point<f32>,
+        r1: f32,
+        color_stops: &'a [ColorStop],
+        extend: u8,
+    },
+    SweepGradient {
+        c0: Point<f32>,
+        start_angle: f32,
+        end_angle: f32,
+        color_stops: &'a [ColorStop],
+        extend: u8,
+    },
+}
+
+pub enum ColrGlyphDrawResult {
+    ColrGlyphDrawn,
+    ColrGlyphDrawingNotImpl,
+}
+// Operates in font units, client needs to apply scale transform.
+pub trait ColorPainter {
+    fn push_transform(&mut self, transform: Transform);
+    fn pop_transform(&mut self);
+
+    fn push_clip_glyph(&mut self, glyph: GlyphId);
+    // Needed for pushing clip box when starting a new root glyph or PaintColrGlyph.
+    fn push_clip_rectangle(&mut self, x_min: f32, x_max: f32, y_min: f32, y_max: f32);
+    fn pop_clip(&mut self);
+
+    fn fill(&mut self, brush: Brush);
+
+    // TODO(drott): Optimization function combining clip and fill.
+    fn draw_colr_glyph(&mut self, _glyph: GlyphId) -> Result<ColrGlyphDrawResult, ReadError> {
+        Ok(ColrGlyphDrawResult::ColrGlyphDrawingNotImpl)
+    }
+
+    fn push_layer(&mut self, composite_mode: CompositeMode);
+    fn pop_layer(&mut self);
+}
+
+pub fn traverse_color_outline<'a>(
+    font: &impl TableProvider<'a>,
+    glyph_id: GlyphId,
+    instance: LocationRef,
+    painter: &mut impl ColorPainter,
+) -> Result<(), ReadError> {
+    let colr = font.colr()?;
+
+    let color_instance = ColrInstance::new(colr, instance.coords());
+    let (root_paint, root_paint_id) =
+        (*color_instance)
+            .v1_base_glyph(glyph_id)?
+            .ok_or(ReadError::MalformedData(
```

Comment:
It’s difficult to find good options for high level errors in `ReadError`. We should probably define a new error type specifically for color.